### PR TITLE
Alerting: Store alertmanager configuration history in a separate table in the database

### LINKED ISSUE #59845: Alerting: Alertmanager configuration truncation in MySQL 8
**What happened**:

Many of the Alertmanager resources are stored in a single JSON blob in the database, as a single string field.

When running against MySQL 8.0.28, the JSON blob will be truncated once it reaches a certain size (roughly 60-100KB). This completely breaks the alerting config.

This bug does not reproduce against MySQL 5.7.40 with the same configuration, in the same environment.

The specific size is not clear. We've seen it truncate at different points on different deployments at the same MySQL version. One case truncated at 65KB, another truncated at 108KB.

The field in question is of type `MEDIUMTEXT`, which according to MySQL docs [should not truncate until 16MB](https://dev.mysql.com/doc/refman/8.0/en/storage-requirements.html#data-types-storage-reqs-strings). In practice it truncates far earlier, only on MySQL 8.

**What you expected to happen**:

I expected the alerting configuration to not be truncated.

**How to reproduce it (as minimally and precisely as possible)**:

1. Set up Grafana against MySQL 8.0.28. I used AWS RDS to do this quickly.
2. Create a lot of alerting objects. In my case, tried to create 1000 contact points.
3. The alertmanager config becomes broken after the first 300 or so, with an error like `unexpected end of JSON value`

**Anything else we need to know?**:

**Environment**:
- Grafana version: main
- Data source type & version: n/a
- OS Grafana is installed on: Linux
- User OS & Browser: n/a
- Grafana plugins: n/a
- Others: MySQL 8.0.28

### PR DESCRIPTION
**Pull request by alexweav**
**What is this feature?**

Instead of storing all (both active and historical) configurations in the same table, we now store the active configurations in a separate table from the historical ones.

**Why do we need this feature?**

This drastically simplifies the SQL required to manage configurations. This fixes a couple existing bugs, and would have prevented others from happening.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/59845
Contributes https://github.com/grafana/grafana/issues/60053

**Special notes for your reviewer**:
n/a

### CODE REVIEW

File: pkg/services/sqlstore/migrations/ualert/tables.go [hunk 1]
```diff
@@ -23,6 +23,8 @@ func AddTablesMigrations(mg *migrator.Migrator) {
 	AddAlertRuleMigrations(mg, 60)
 	AddAlertRuleVersionMigrations(mg)
 
+	AddAlertmanagerConfigHistoryMigrations(mg)
```
Comment (alexweav) [inline pkg/services/sqlstore/migrations/ualert/tables.go:26]: This is the only safe place to put this. Even then, it's prone to bugs. See https://github.com/grafana/grafana/issues/60196

**Review by gotjosh — APPROVED** [review]

**Comment by armandgrillet** [comment]
```
--- FAIL: Test_subscribeToFolderChanges (1.02s)
    ngalert_test.go:62: 
        	Error Trace:	/drone/src/pkg/services/ngalert/ngalert_test.go:62
        	Error:      	Condition never satisfied
        	Test:       	Test_subscribeToFolderChanges
        	Messages:   	scheduler was expected to be called 5 times but called 0
FAIL
```

**Comment by alexweav** [comment]
@armandgrillet yep, this is an intermittent test failure that is not related to or caused by this change. a huge percentage of valid builds are blocked by intermittent test failures lately, it's something we have to [work](https://github.com/grafana/grafana/issues/60132) [around](https://github.com/grafana/grafana/issues/60320) [repeatedly](https://github.com/grafana/grafana/issues/60107).

**Comment by alexweav** [comment]
Found an edge case in SQLite that doesn't work because of its dynamic typing:
https://www.sqlite.org/faq.html#q3

**Comment by alexweav** [comment]
Fixed, resetting.

**Comment by alexweav** [comment]
The migration was tested on both MySQL 8 and SQLite, with a number of scenarios:
- Multiple orgs, with multiple config histories per org
- Configurations for orgs that don't exist (correctly preserved)
- Very long alertmanager configurations, to confirm they aren't truncated when moved
- Empty configuration table
- Fresh install of Grafana

File: pkg/services/ngalert/store/alertmanager.go [hunk 2]
```diff
@@ -78,10 +74,23 @@ func (st DBstore) SaveAlertmanagerConfigurationWithCallback(ctx context.Context,
 			ConfigurationVersion:      cmd.ConfigurationVersion,
 			Default:                   cmd.Default,
 			OrgID:                     cmd.OrgID,
+			CreatedAt:                 time.Now().Unix(),
```
Comment (gotjosh) [inline pkg/services/ngalert/store/alertmanager.go:77]: why do we need to do this? This column _should_ be automanaged by xorm?

File: pkg/services/sqlstore/migrations/ualert/tables.go [hunk 3]
```diff
@@ -345,6 +345,16 @@ func AddAlertmanagerConfigMigrations(mg *migrator.Migrator) {
...
+
+	ExtractAlertmanagerConfigurationHistoryMigration(mg)
+
+	mg.AddMigration("drop non-unique orgID index", migrator.NewDropIndexMigration(alertConfiguration, &migrator.Index{
+		Cols: []string{"org_id"},
+	}))
+	mg.AddMigration("add unique index on orgID", migrator.NewAddIndexMigration(alertConfiguration, &migrator.Index{
+		Type: migrator.UniqueIndex,
+		Cols: []string{"org_id"},
+	}))
```
Comment (gotjosh) [inline pkg/services/sqlstore/migrations/ualert/tables.go:357]: Had a discussion offline about https://github.com/grafana/grafana/issues/60196. Can we move this to L38 to make sure we don't need fueling the same problem?

**Review by gotjosh — APPROVED** [review]
LGTM - my comments are nits and you can issue them to tackle them later on.

File: pkg/services/ngalert/store/alertmanager.go (hunk 2 above)
Comment (alexweav) [reply pkg/services/ngalert/store/alertmanager.go:77]: Will address in the future as discussed offline

**Pull request merged**